### PR TITLE
Reduce network pressure on multi row insert

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR adds bind variables pruning to only required bind variables per shard. This reduce the network overhead of sending non-required bind variable values.
This becomes evident when 100's of shards are involved in a bulk insert query.

I added a benchmark that spawns 16 shards and insert 10k rows as multi-row insert.
The table contains 13 columns with per row size of ~600 bytes.
The new implementation gives an execution improvement by ~70% and reduce memory utilization by ~60%

```
pkg: vitess.io/vitess/go/test/endtoend/vtgate/queries/benchmark
                      │  main-2.out  │            insert-2.out             │
                      │    sec/op    │   sec/op     vs base                │
ShardedTblNoLookup-10   1134.3m ± 8%   345.6m ± 7%  -69.53% (p=0.000 n=10)

                      │  main-2.out   │             insert-2.out              │
                      │     B/op      │     B/op       vs base                │
ShardedTblNoLookup-10   71.92Mi ± 40%   28.07Mi ± 19%  -60.97% (p=0.000 n=10)

                      │  main-2.out  │             insert-2.out             │
                      │  allocs/op   │  allocs/op    vs base                │
ShardedTblNoLookup-10   680.0k ± 50%   226.7k ± 25%  -66.67% (p=0.000 n=10)
```

Note: p-value of 0.000 indicates `extremely unlikely - difference between the medians is due to chance`
https://pkg.go.dev/golang.org/x/perf/cmd/benchstat

On running the same benchmark with 32 shards, it is just the CPU contention increases on the existing implementation
```
pkg: vitess.io/vitess/go/test/endtoend/vtgate/queries/benchmark
                      │ main-32-shards.out │        insert-32-shards.out         │
                      │       sec/op       │   sec/op     vs base                │
ShardedTblNoLookup-10        2090.3m ± 24%   365.0m ± 7%  -82.54% (p=0.000 n=10)

                      │ main-32-shards.out │         insert-32-shards.out          │
                      │        B/op        │     B/op       vs base                │
ShardedTblNoLookup-10         71.93Mi ± 0%   28.06Mi ± 12%  -60.99% (p=0.000 n=10)

                      │ main-32-shards.out │         insert-32-shards.out         │
                      │     allocs/op      │  allocs/op    vs base                │
ShardedTblNoLookup-10          680.0k ± 0%   226.7k ± 25%  -66.67% (p=0.000 n=10)
```

On running the same benchmark with 8 shards, the contention reduces but the difference is still significant.
```
pkg: vitess.io/vitess/go/test/endtoend/vtgate/queries/benchmark
                      │ main-8-shards.out │         insert-8-shards.out         │
                      │      sec/op       │   sec/op     vs base                │
ShardedTblNoLookup-10         608.2m ± 6%   341.5m ± 7%  -43.86% (p=0.000 n=10)

                      │ main-8-shards.out │          insert-8-shards.out          │
                      │       B/op        │     B/op       vs base                │
ShardedTblNoLookup-10        43.02Mi ± 9%   26.33Mi ± 17%  -38.80% (p=0.000 n=10)

                      │ main-8-shards.out │         insert-8-shards.out          │
                      │     allocs/op     │  allocs/op    vs base                │
ShardedTblNoLookup-10         340.0k ± 0%   198.4k ± 14%  -41.67% (p=0.000 n=10)
```

Ran 16 shards benchmark with different numbers of rows. Evidently, `CPU utilization drops` significantly with the new splitting logic and sending only required bind variables per shard.

```
pkg: vitess.io/vitess/go/test/endtoend/vtgate/queries/benchmark
                                           │ main-16-shards.out │         insert-16-shards.out         │
                                           │       sec/op       │    sec/op     vs base                │
ShardedTblNoLookup/16-shards-1-rows-10             348.4µ ±  6%   345.5µ ±  3%        ~ (p=0.739 n=10)
ShardedTblNoLookup/16-shards-10-rows-10            4.307m ±  6%   3.630m ± 11%  -15.74% (p=0.000 n=10)
ShardedTblNoLookup/16-shards-100-rows-10           17.52m ±  7%   11.07m ±  8%  -36.83% (p=0.000 n=10)
ShardedTblNoLookup/16-shards-500-rows-10           52.79m ±  3%   25.26m ±  9%  -52.15% (p=0.000 n=10)
ShardedTblNoLookup/16-shards-1000-rows-10         105.05m ±  5%   36.93m ±  3%  -64.84% (p=0.000 n=10)
ShardedTblNoLookup/16-shards-5000-rows-10          520.9m ±  5%   151.2m ±  9%  -70.96% (p=0.000 n=10)
ShardedTblNoLookup/16-shards-10000-rows-10        1113.6m ± 20%   288.8m ±  4%  -74.07% (p=0.000 n=10)
geomean                                            36.40m         17.80m        -51.10%

                                           │ main-16-shards.out │         insert-16-shards.out         │
                                           │        B/op        │     B/op       vs base               │
ShardedTblNoLookup/16-shards-1-rows-10             1000.0 ±  0%    1000.0 ±  0%       ~ (p=0.581 n=10)
ShardedTblNoLookup/16-shards-10-rows-10           6.602Ki ±  1%   6.601Ki ±  1%       ~ (p=0.105 n=10)
ShardedTblNoLookup/16-shards-100-rows-10          66.33Ki ±  1%   66.20Ki ±  1%       ~ (p=0.954 n=10)
ShardedTblNoLookup/16-shards-500-rows-10          364.1Ki ±  5%   351.1Ki ±  4%       ~ (p=0.061 n=10)
ShardedTblNoLookup/16-shards-1000-rows-10         763.9Ki ± 17%   697.1Ki ± 11%       ~ (p=0.305 n=10)
ShardedTblNoLookup/16-shards-5000-rows-10         4.748Mi ± 35%   4.146Mi ± 16%       ~ (p=0.304 n=10)
ShardedTblNoLookup/16-shards-10000-rows-10        10.15Mi ± 39%   10.14Mi ± 20%  -0.08% (p=0.021 n=10)
geomean                                           179.5Ki         172.8Ki        -3.74%

                                           │ main-16-shards.out │         insert-16-shards.out          │
                                           │     allocs/op      │  allocs/op   vs base                  │
ShardedTblNoLookup/16-shards-1-rows-10              6.000 ±  0%   6.000 ±  0%        ~ (p=1.000 n=10) ¹
ShardedTblNoLookup/16-shards-10-rows-10             6.000 ±  0%   6.000 ±  0%        ~ (p=1.000 n=10) ¹
ShardedTblNoLookup/16-shards-100-rows-10            6.000 ±  0%   6.000 ±  0%        ~ (p=1.000 n=10) ¹
ShardedTblNoLookup/16-shards-500-rows-10            6.000 ±  0%   6.000 ±  0%        ~ (p=1.000 n=10) ¹
ShardedTblNoLookup/16-shards-1000-rows-10           6.500 ±  8%   6.000 ±  0%        ~ (p=0.141 n=10)
ShardedTblNoLookup/16-shards-5000-rows-10          10.000 ± 20%   8.000 ± 12%  -20.00% (p=0.005 n=10)
ShardedTblNoLookup/16-shards-10000-rows-10          13.00 ± 23%   10.00 ± 20%  -23.08% (p=0.003 n=10)
```

## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/14077

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [X] Tests were added or are not required
-   [X] Did the new or modified tests pass consistently locally and on the CI
-   [X] Documentation was added or is not required

### CODE REVIEW

**Review by vmg — APPROVED** [review]
Looks good! Nice fix!

File: go/test/endtoend/vtgate/queries/benchmark/benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2022 The Vitess Authors.
```
Comment (GuptaManan100) [inline go/test/endtoend/vtgate/queries/benchmark/benchmark_test.go:2]: 2023 is the year you're looking for

File: go/test/endtoend/vtgate/queries/benchmark/benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		allRows = append(allRows, "("+strings.Join(row, ",")+")")
+	}
+	return fmt.Sprintf("insert into %s(%s) values %s", tq.tableName, strings.Join(tq.cols, ","), strings.Join(allRows, ","))
+}
+
+func getRandomString(size int) string {
+	var str strings.Builder
+
+	for i := 0; i < size; i++ {
+		str.WriteByte(byte((rand.Int() % 26) + 97))
```
Comment (GuptaManan100) [inline go/test/endtoend/vtgate/queries/benchmark/benchmark_test.go:55]: `rand.Int() % 26` can just be replaced with `rand.Intn(27)`

File: go/test/endtoend/vtgate/queries/benchmark/main_test.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2022 The Vitess Authors.
```
Comment (GuptaManan100) [inline go/test/endtoend/vtgate/queries/benchmark/main_test.go:2]: 2023

**Review by GuptaManan100 — APPROVED** [review]
Love it 🚀

**Pull request merged**